### PR TITLE
chore: extract load data into table as sub workflow

### PR DESCRIPTION
**Pull request by dengmingtong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

Split load workflow

## Implementation highlights

1. Split load workflow
2. Add UT

## Test checklist

- [ ] add new test cases
- [x] all code changes are covered by unit tests
- [x] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [x] deploy data modeling
    - [x] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/analytics/lambdas/load-data-workflow/has-more-job-new.ts [hunk 1]
```diff
@@ -22,24 +22,26 @@ const ODS_EVENT_BUCKET = process.env.ODS_EVENT_BUCKET!;
 const ODS_EVENT_BUCKET_PREFIX = process.env.ODS_EVENT_BUCKET_PREFIX!;
 const DYNAMODB_TABLE_NAME = process.env.DYNAMODB_TABLE_NAME!;
 const DYNAMODB_TABLE_INDEX_NAME = process.env.DYNAMODB_TABLE_INDEX_NAME!;
-const REDSHIFT_ODS_TABLE_NAME = process.env.REDSHIFT_ODS_TABLE_NAME!;
 
 export const handler = async (_: any, context: Context) => {
```
Comment (zxkane) [inline src/analytics/lambdas/load-data-workflow/has-more-job-new.ts:26]: rename `_` to `event`, and use the type

same for other handlers

File: src/analytics/private/load-ods-data-workflow.ts [hunk 2]
```diff
@@ -177,15 +180,138 @@ export class LoadOdsDataToRedshiftWorkflow extends Construct {
...
+    const logGroup = createLogGroup(this,
+      {
+        prefix: `/aws/vendedlogs/states/Clickstream/LoadData-${this.node.id}`,
+      },
+    );
+
+    let parallelLoadDataToTables = new Parallel(this, `${this.node.id} - Load data to tables`);
+    const subWorkflow = this.createSubWorkflow(ddbTable, props, copyRole, logGroup);
+    for (const odsTable of Object.keys(props.tablesOdsSource)) {
+      const subExecution = new StepFunctionsStartExecution(this, `${this.node.id} - ${odsTable}`, {
```
Comment (zxkane) [inline src/analytics/private/load-ods-data-workflow.ts:245]: specify the name, like the below pattern for easy debugging,

`<parent execution id>-<table name>`

File: src/analytics/private/load-ods-data-workflow.ts [hunk 3]
```diff
@@ -177,15 +180,138 @@ export class LoadOdsDataToRedshiftWorkflow extends Construct {
...
+        }),
+      ],
+    }));
+
+    return loadDataStateMachine;
+  }
+
+  private createSubWorkflow(
+    ddbTable: ITable,
+    // odsTableName: string,
```
Comment (zxkane) [inline src/analytics/private/load-ods-data-workflow.ts:301]: remove it

File: src/analytics/analytics-on-redshift.ts [hunk 4]
```diff
@@ -304,6 +307,8 @@ export class RedshiftAnalyticsStack extends NestedStack {
       tablesOdsSource: props.tablesOdsSource,
       workflowBucketInfo: props.workflowBucketInfo,
       loadDataConfig: props.loadDataConfig,
+      odsSourceS3Bucket: props.dataSourceBucket,
```
Comment (zxkane) [inline src/analytics/analytics-on-redshift.ts:310]: Is it really needed? We have bucket and prefix info in type `TablesODSSource`.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/analytics/private/load-ods-data-workflow.ts (hunk 3 above)
Comment (dengmingtong) [reply src/analytics/private/load-ods-data-workflow.ts]: done

File: src/analytics/analytics-on-redshift.ts (hunk 4 above)
Comment (dengmingtong) [reply src/analytics/analytics-on-redshift.ts:310]: change bucket and prefix to input parameter of step function workflows

File: src/analytics/private/load-ods-data-workflow.ts (hunk 2 above)
Comment (dengmingtong) [reply src/analytics/private/load-ods-data-workflow.ts]: changed to <parent execution name>-<table name>.
The execution id is too long.

File: src/analytics/lambdas/load-data-workflow/has-more-job-new.ts (hunk 1 above)
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/has-more-job-new.ts]: Refactored them.

File: src/analytics/private/load-ods-data-workflow.ts [hunk 5]
```diff
@@ -177,15 +176,142 @@ export class LoadOdsDataToRedshiftWorkflow extends Construct {
...
+    // Next state machines
+    //
+    let nexTaskExecChain;
+    for (const nexTask of props.nextStateStateMachines) {
+      const taskName = nexTask.name;
+      const stateMachine = nexTask.stateMachine;
+      const input = nexTask.input;
+      const integrationPattern = nexTask.integrationPattern?? IntegrationPattern.REQUEST_RESPONSE;
+      const resultPath = `$.${nexTask.resultPath ?? taskName.replace(/ /g, '')}`;
+      const nexTaskExec = new StepFunctionsStartExecution(this, `${this.node.id} - ${taskName}`, {
```
Comment (zxkane) [inline src/analytics/private/load-ods-data-workflow.ts:206]: specify below parameters to specify the execution name of sub workflow,

```json
{
  "StateMachineArn": "state arn",
  "Input": {
    "AWS_STEP_FUNCTIONS_STARTED_BY_EXECUTION_ID.$": "$$.Execution.Id"
  },
  "Name.$": "States.Format('{}-{}', States.ArrayGetItem(States.StringSplit($$.Execution.Id, ':'), 7), 'my-suffix')"
}
```

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/analytics/lambdas/load-data-workflow/create-load-manifest.ts [hunk 6]
```diff
@@ -240,15 +280,21 @@ function getAppIdFromS3Object(s3Object: string) {
...
  */
 
-export const queryItems = async (tableName: string, indexName: string, prefix: string, jobStatus: string, lastEvaluatedKey: any): Promise<{
+export const queryItems = async (
+  tableName: string,
+  indexName: string,
+  prefix: string,
+  jobStatus: string,
+  lastEvaluatedKey: any,
+  odsTableName?: string,
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/create-load-manifest.ts:289]: https://github.com/awslabs/clickstream-analytics-on-aws/blob/main/src/analytics/lambdas/load-data-workflow/skip-running-workflow.ts#L136

File: src/analytics/lambdas/load-data-workflow/load-manifest-to-redshift.ts [hunk 7]
```diff
@@ -51,6 +51,9 @@ type LoadManifestEventDetail = ManifestBody & {
 
 export interface LoadManifestEvent {
   detail: LoadManifestEventDetail;
+  odsTableName: string;
+  odsSourceBucket: string;
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/load-manifest-to-redshift.ts]: Fixed

File: src/analytics/lambdas/load-data-workflow/has-more-job-new.ts [hunk 8]
```diff
@@ -81,6 +90,9 @@ export const handler = async (_: any, context: Context) => {
     processingFilesCount: tableProcessingCountInfo,
     jobNewCount: currentJobNewCount,
     hasMoreWork: (currentJobNewCount + totalProcessCount) > 0,
+    odsTableName: odsTableName,
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/has-more-job-new.ts]: Fixed

File: src/analytics/lambdas/load-data-workflow/create-load-manifest.ts [hunk 9]
```diff
@@ -214,6 +253,7 @@ const doManifestFiles = async (candidateItems: Array<ODSEventItem>,
   return {
     manifestList: manifestFiles,
     count: manifestFiles.length,
+    odsTableName: odsTableName,
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/create-load-manifest.ts]: Fixed

File: src/analytics/lambdas/load-data-workflow/create-load-manifest.ts [hunk 10]
```diff
@@ -161,10 +196,14 @@ export const handler = async (_event: any, context: Context) => {
   logger.info('FILE_NEW=' + allJobNewCount + ', FILE_PROCESSING=' + processInfo.jobNum
     + ', FILE_ENQUEUE=' + enQueueInfo.jobNum + ', FILE_MAX_AGE=' + maxFileAgeSeconds);
 
-  return response;
+  return {
+    ...response,
+    odsSourceBucket: odsSourceBucket,
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/create-load-manifest.ts]: Fixed

File: src/analytics/lambdas/load-data-workflow/check-load-status.ts [hunk 11]
```diff
@@ -49,6 +49,9 @@ type CheckLoadStatusEventDetail = ManifestBody & {
 export interface CheckLoadStatusEvent {
   detail: CheckLoadStatusEventDetail;
   waitTimeInfo: WaitTimeInfo;
+  odsTableName: string;
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/load-data-workflow/check-load-status.ts]: Fixed

File: src/analytics/analytics-on-redshift.ts (hunk 4 above)
Comment (dengmingtong) [reply src/analytics/analytics-on-redshift.ts:310]: example:
  const tablesOdsSource: TablesODSSource = {
    event: {
      s3Bucket: props.dataSourceConfiguration.bucket,
      prefix: props.dataSourceConfiguration.prefix + TABLE_NAME_EVENT + '/',
      fileSuffix: props.dataSourceConfiguration.fileSuffix,
    },
    event_parameter: {
      s3Bucket: props.dataSourceConfiguration.bucket,
      prefix: props.dataSourceConfiguration.prefix + TABLE_NAME_EVENT_PARAMETER + '/',
      fileSuffix: props.dataSourceConfiguration.fileSuffix,
    },
    user: {
      s3Bucket: props.dataSourceConfiguration.bucket,
      prefix: props.dataSourceConfiguration.prefix + TABLE_NAME_USER + '/',
      fileSuffix: props.dataSourceConfiguration.fileSuffix,
    },
    item: {
      s3Bucket: props.dataSourceConfiguration.bucket,
      prefix: props.dataSourceConfiguration.prefix + TABLE_NAME_ITEM + '/',
      fileSuffix: props.dataSourceConfiguration.fileSuffix,
    },
  };

File: src/analytics/private/load-ods-data-workflow.ts (hunk 5 above)
Comment (dengmingtong) [reply src/analytics/private/load-ods-data-workflow.ts:206]: Fixed

File: src/data-analytics-redshift-stack.ts [hunk 12]
```diff
@@ -137,6 +137,8 @@ export function createRedshiftAnalyticsStack(
     loadDataConfig,
     workflowBucketInfo,
     mvRefreshInterval: props.redshift.mvRefreshInterval,
+    dataSourceBucket: props.dataSourceConfiguration.bucket,
```
Comment (zxkane) [inline src/data-analytics-redshift-stack.ts:140]: remove it if it's useless

File: src/analytics/analytics-on-redshift.ts [hunk 13]
```diff
@@ -64,6 +65,8 @@ export interface RedshiftAnalyticsStackProps extends NestedStackProps {
   readonly clearExpiredEventsWorkflowData: ClearExpiredEventsWorkflowData;
   readonly emrServerlessApplicationId: string;
   readonly dataProcessingCronOrRateExpression: string;
+  readonly dataSourceBucket: IBucket;
```
Comment (zxkane) [inline src/analytics/analytics-on-redshift.ts:68]: remove it if it's uesless

**Review by zxkane — CHANGES_REQUESTED** [review]

File: test/analytics/analytics-on-redshift/workflow/load-data-stepfuncs.test.ts [hunk 14]
```diff
@@ -38,5 +39,19 @@ describe('Load data workflow', () => {
     expect(JSON.stringify(strCapture.asObject(), undefined, 2)).toEqual(loadDataStepFuncDef);
   });
```
Comment (zxkane) [inline test/analytics/analytics-on-redshift/workflow/load-data-stepfuncs.test.ts:41]: add test case to Verify the sub-workflow is invoked with proper properties to map the tables

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**